Sources/JS/helpers.js

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+var maxCollectedProperties = 200
+// groupName: constructor
+var possibleGroups = [
+    String, RegExp, Array, Map, Promise, WeakMap, Set, WeakSet, ArrayBuffer, DataView, Uint8Array, Int8Array,
+    Uint16Array, Int16Array, Uint32Array, Int32Array, Float32Array, Float64Array, Uint8ClampedArray, Object
+]
+// Must be kept in sync with TypeSystem.swift
+baseTypes = {
+    nothing: 0,
+    undefined: 1 << 0,
+    integer: 1 << 1,
+    float: 1 << 2,
+    string: 1 << 3,
+    boolean: 1 << 4,
+    object: 1 << 5,
+    function: 1 << 6,
+    constructor: 1 << 7,
+    unknown: 1 << 8,
+    bigint: 1 << 9,
+    regexp: 1 << 10,
+}
+function setGroup(value, currentType) {
+    for (var i=0;i<possibleGroups.length;i++) {
+        try {
+            if (value instanceof possibleGroups[i]) {
+                currentType.group = possibleGroups[i].name
```

Comment:
Just flagging a potential gotcha: while doing `String.name = "foo"` isn't possible (`.name` is read-only), script code could do `delete String.name` in which case this code here wouldn't work correctly anymore. Since this is an edge-case, I'd be fine with just mentioning this in a comment.

Reply:
Well I think maybe more general mechanism can be used. Instead of array of constructor we define list of groups as pairs (`groupName`, `belongsToGroup`), where belongsToGroup is function which takes object and returns boolean if object belongs to particular group. So we don't rely on having property `name` or that group has to be checked by `instanceof`

Reply:
Yup, that also SGTM!